### ISSUE TITLE
Extend mappings of `createUnistylesComponent` to support dynamic styling

### ISSUE THREAD

**Issue by jpudysz** [jpudysz/react-native-unistyles#362]
### Description

This question was reported on Bluesky, where Mazen asked about a migration guide for this case:

![3cf155d0-049f-4704-a12a-c106eeb8cecd](https://github.com/user-attachments/assets/bfc7df23-4f98-4ce7-9270-cb026680abd2)

In this scenario, wrapping `Switch` in `createUnistylesComponent` won't work because both `thumbColor` and `trackColor` use different colors based on `prop.disabled`.

I believe it might be doable to enhance `createUnistylesComponent` to pass a second parameter:

**Before:**
```ts
const UnistyleSwitch = createUnistylesComponent(Switch, theme => ({
    // 😭 We can't do much as we don't have access to the props
}))
```

**After:**
```ts
const UnistyleSwitch = createUnistylesComponent(Switch, (theme, props) => ({
    // 😄 Combining theme and props is possible now
}))
```

I need to take some time to think it through and come up with a solution for the next beta release.

### Steps to Reproduce

Make this work based on the example from above.

### Unistyles Version

3.0.0-beta.1

### React Native Version

0.76.3

### Platforms

Android, iOS, Web

### Expo

Yes

**Comment by xinha-sh** [comment]
I tried migrating my app to 3.0 and most of the API decisions seemed very intuitive but this`createUnistylesComponent ` api leads to a major refactor. I change fill colour of svg icons using theme and have used theme.x.y at various places. 

I think this `createUnistylesComponent` API ensure no-renders. But for 3.0, a escape hatch should be provided (so that gradually this API can be adopted to codebase). 

Maybe a hook `const theme = useTheme()` or  `useStyles` with deprecation warnings or a function that can be called outside a component like 
```ts
const theme = getCurrentTheme()
export default function Index() {
 ....
  return <Menu
    size={28}
     color={isDarkColorScheme ? theme.colors.brand25 : tintColor}
   />
}
```

**Comment by jpudysz** [comment]
Here is another example to consider (from Discord):

```tsx
export const TitleSelectSection = memo(() => {
  const {styles} = useStyles(stylesheet);
  const columnCount = useBreakpoints({xs: 2, md: 3});

  const {itemList, fetchMore} = useTitles();
  const {updateTitle, currentTitleId} = useTitleControl();

  return (
    <VStack style={styles.root}>
      <FlashList
        numColumns={columnCount} // <-- depends on breakpoint / custom props
        ...truncated/>
    </VStack>
  );
});
```

**Comment by thecynicalpaul** [comment]
> Then there is no point of any optimization. Using such hook on the root level of your app will re-render everything 😢. You might use it in your icon component, but someone will definitely use it on the root level and will complain "where are my no re-renders!?".

There is an argument to be made that wrapping a component with `createUnistylesComponent` can create the exact same scenario, except instead of a hook, you have a decorator. 😋

**Comment by thecynicalpaul** [comment]
I, personally, would love to see an API similar to reanimated, where we'd have two hooks:
`useDerivedUnistyles` (-> `useDerivedValue`, [ref](https://docs.swmansion.com/react-native-reanimated/docs/core/useDerivedValue/))
`useUnistylesReaction` (-> `useAnimatedReaction`, [ref](https://docs.swmansion.com/react-native-reanimated/docs/advanced/useAnimatedReaction))

Something like:
```tsx
const someUnistylesHybridObject = useDerivedUnistyles((theme, rt) => {
  return rt.breakpoints.xs; // or whatever
}, [...deps]);
```

This way we could also shoehorn our own re-rendering system via `useState`, which would preserve the "non-re-rendering purity" of Unistyles. For example:
```tsx
const [state, setState] = useState<number | null>(null);

const someUnistylesHybridObject = useUnistylesReaction((theme, rt) => rt.breakpoint, breakpoint => {
  if (breakpoint === "xs") {
    setState(MAGIC_CONST_0);
    return;
  }

  setState(MAGIC_CONST_1);
}, [setState]);
```

**Comment by jpudysz** [comment]
Alright, everyone, here’s our proposal for what will be implemented in `beta.2`. We brainstormed extensively and tried to find the best solution based on your examples.

---

### 1. Deprecate `createUnistylesComponent`

Once you install `beta.2`, you’ll see a `deprecation` warning indicating that `createUnistylesComponent` is no longer the recommended approach. We believe a shorter version will work better, so from now on, we’ll use `withUnistyles`.

**Example:**

```ts
const UniButton = withUnistyles(Button)

export const App = () => { ... }
```

---

### 2. `style` and `contentContainerStyles`

As before, if any third-party component requires updates for these props, we will handle it out of the box.

**Example:**

```ts
const UniFlashList = withUnistyles(FlashList)
```

---

### 3. Global Mappings

If you have a simple example and want to share the same props across all your components, you can still do so as before:

**Example:**

```ts
// your button will always be tied to `button` color from theme
const UniButton = withUnistyles(Button, theme => ({
    color: theme.colors.button
}))
```

---

### 4. `miniRuntime`

We will also provide `miniRuntime` for you:

**Example:**

```ts
const UniButton = withUnistyles(Button, (theme, rt) => ({
    ...
}))
```

---

### 5. Advanced Use Cases with `uniProps`

If you want to access other props, variables from a closure, or do something more dynamic, you can use `uniProps`:

**Example:**

```ts
const UniCustomView = withUnistyles(CustomView)

export const App = () => {
   const externalValue = 3
   const externalFunction = (index, theme) => {
       return externalValue === index + externalValue
          ? theme.colors.primary
          : theme.colors.secondary
   }
    
   return (
        <UniCustomView
              uniProps={(theme, rt) => ({
                    prop1: theme.colors.background,
                    renderListItem: ({ item, index }) => {
                          const color = externalFunction(index, theme)
                          
                          return (
                              ...
                          )
                    }
              })}
        />
   )
}
```

With `uniProps`, you can pass a function, modify props, etc., while having access to both `theme` and `rt`.  

---

### Prop Resolution Priority

We will respect your order of prop resolution, applying them with the following priority:

1. **Global mappings**
2. **`uniProps`**
3. **Inline props**

**Example: Modifying a Button**

```ts
// By default, the button is red.
const UniButton = withUnistyles(Button, theme => ({
    color: theme.colors.red
}))

// `themedProps` have higher priority,
// so the button is orange.
<UniButton
    uniProps={theme => ({
       color: theme.colors.orange
    })}
/>

// Inline props have the highest priority,
// so the button is pink.
<UniButton
    uniProps={theme => ({
       color: theme.colors.orange
    })}
    color="pink"
/>
```

---

We hope this resolves all blockers and is easy to adopt across custom components. Of course, we don’t wrap your component in anything. We simply force re-rendering for such components and return them as-is. We also read your component dependencies from `StyleSheet` and re-render only when needed.

Here are examples based on your snippets:

```tsx
import { withUnistyles } from 'react-native-unistyles'
import { Switch } from 'react-native'

interface ToggleProps extends SwitchProps {}

const UniSwitch = withUnistyles(Switch)

const ToggleExample = ({ ...rest }: ToggleProps) => (
    <UniSwitch
         uniProps={theme => ({
              thumbColor: rest.disabled && !rest.value ? theme.palette.grey[50] : theme.palette.primary.white,
              trackColor: {
                  false: rest.disabled ? theme.palette.grey[100] : theme.palette.grey[50],
                  true: rest.disabled ? theme.palette.purple[50] : theme.palette.accent.purple
              }
         })}
    />
)
```

```tsx
const UniList = withUnistyles(List)

export default function HomeScreen() {
    const router = useRouter();

  const getColorFromValue = (value: number, theme: UnistylesTheme) => {
    if (value > 0) return theme.colors.mainGreen;
    if (value < 0) return theme.colors.mainRed;

    return theme.colors.mainAppYellow;
  };

  return (
    <ScrollView style={styles.container}>
      <UniList
        data={mockTransactions}
        keyExtractor={(item, index) => index.toString()}
        uniProps={theme => ({
            renderItem={({ item, index }) => (
              <ListItem color={getColorFromValue(item.value, theme)}>
                <ListItem.Label>{item.label}</ListItem.Label>
                <ListItem.Value>{item.value}</ListItem.Value>
              </ListItem>
            )}
        })}
      />
    </ScrollView>
  );
}

const styles = StyleSheet.create((theme) => ({
  container: {
    flex: 1,
    padding: 16,
    backgroundColor: theme.colors.background,
  },
  text: {
    color: '#fff',
  },
}));
```

```tsx
const UniFlashList = withUnistyles(FlashList)

export const TitleSelectSection = memo(() => {
  const getColumnCount = (theme, rt) => {
      // access to breakpoint with rt.breakpoint
     // return column count
  }

  const {itemList, fetchMore} = useTitles();
  const {updateTitle, currentTitleId} = useTitleControl();

  return (
    <VStack style={styles.root}>
      <UniFlashList
        uniProps={(theme, rt) => ({
             numColumns={getColumnCount(theme, rt)}
        })}
        ...truncated/>
    </VStack>
  );
});
```

Keep in mind that we don't yet support reading values in `worklets`, so reading theme in `useAnimatedStyles` etc. for give you updates about theme change.
